parlai/tasks/cmu_dog/agents.py

```diff
@@ -0,0 +1,370 @@
+        if knowledge_keys and k not in knowledge_keys:
+            continue
+        fact = f"{k}:"
+        if isinstance(v, str):
+            fact += v
+        else:
+            fact += ",".join(v)
+        texts.append(fact)
+    return fact_delimiter.join(texts)
+
+
+def _build_rare_word_f1(opt: Opt) -> RareWordF1Calculator:
+    datapath = _datapath(opt)
+
+    def _collect_convo_text(convo_data):
+        convo_texts = []
+        for conv in convo_data.values():
+            # get all messages
+            convo_texts.append(' '.join([m['text'] for m in conv['history']]))
+        return convo_texts
+
+    # use conversation data from all splits for consistency
+    convos = []
+    convo_files = _all_split_datafiles(opt)
+
+    for fname in convo_files:
+        with PathManager.open(os.path.join(datapath, f"conversations/{fname}")) as f:
+            data = json.load(f)
+        convos += _collect_convo_text(data)
+
+    return RareWordF1Calculator(' '.join(convos), top_p=0.5)
+
+
+class CMUDocumentGroundedConversationsTeacher(DialogTeacher):
+    """
+    CMU Document Grounded Conversations Dataset (aka CMU_DoG)
+
+    Paper: https://arxiv.org/pdf/1809.07358.pdf
+    Source: https://github.com/festvox/datasets-CMU_DoG
+    """
+
+    def __init__(self, opt: Opt, shared: TShared = None):
+        opt = copy.deepcopy(opt)
+        self.delimiter = opt.get('delimiter', '\n')
+        split_type = SplitType(opt.get("cmu_dog_split_type"))
+        if split_type == SplitType.ORIGINAL:
+            logger.warning(
+                "`original` split type contains duplicate conversations across train, "
+                "valid, and test. See https://github.com/festvox/datasets-CMU_DoG/issues/2 "
+                "for more detail."
+            )
+        opt['datafile'] = _datafile(
+            split=DatatypeHelper.fold(opt['datatype']), split_type=split_type
+        )
+        super().__init__(opt, shared)
+        if shared:
+            self.rare_word_f1 = shared['rare_word_f1']
+        else:
+            self.rare_word_f1 = _build_rare_word_f1(opt)
+
+    @classmethod
+    def add_cmdline_args(
+        cls, parser, partial_opt: Optional[Opt] = None
+    ) -> ParlaiParser:
+        super().add_cmdline_args(parser, partial_opt=partial_opt)
+        cmu_dog = parser.add_argument_group("CMU Document Grounded Conversations")
+        cmu_dog.add_argument(
+            "--cmu-dog-rating",
+            type=int,
+            default=[1, 2, 3],
+            choices=[1, 2, 3],
+            nargs="+",
+            help='The higher the number, the better quality the conversation. '
+            'For each rating, the number of conversations is as follows: 1-1443, 2-2142, 3-527',
+        )
+        cmu_dog.add_argument(
+            "--cmu-dog-only-with-knowledge",
+            type=bool,
+            default=True,
+            help="Optionally train only the sides of the conversation that have access to knowledge.",
```

Comment:
nit: as this defaults to True, and is presumably the desired arg, might be good to update the help comment here, like "Set False to optionally train both sides of conversation"